<launch>
  <arg name="ODOM_FRAME" default="odom" />

  <group ns="laser_preprocess">
    <node pkg="jsk_topic_tools" type="standalone_complexed_nodelet"
          name="laser_preprocess_manager" output="screen" respawn="true">
      <rosparam>
        nodelets:
         - name: downsampler
           type: pcl/VoxelGrid
           remappings:
             - from: ~input
               to: /tilt_laser_listener/output_cloud
         - name: gsensor_cloud
           type: jsk_pcl/TfTransformCloud
           remappings:
             - from: ~input
               to: downsampler/output
         - name: camera_laser_cloud
           type: jsk_pcl/TfTransformCloud
           remappings:
             - from: ~input
               to: downsampler/output
         - name: x_filter
           type: pcl/PassThrough
           remappings:
             - from: ~input
               to: gsensor_cloud/output
         - name: y_filter
           type: pcl/PassThrough
           remappings:
             - from: ~input
               to: x_filter/output
         - name: z_filter
           type: pcl/PassThrough
           remappings:
             - from: ~input
               to: y_filter/output
         - name: odom_cloud
           type: jsk_pcl/TfTransformCloud
           remappings:
             - from: ~input
               to: z_filter/output
      </rosparam>
    </node>
    <group ns="downsampler">
      <rosparam>
        leaf_size: 0.02
        filter_limit_min: -100
        filter_limit_max: 100
        skip_number: 10
      </rosparam>
    </group>
    <group ns="gsensor_cloud">
      <rosparam>
        target_frame_id: ground
      </rosparam>
    </group>
    <group ns="camera_laser_cloud">
      <rosparam>
        target_frame_id: left_camera_optical_frame
      </rosparam>
    </group>

    <group ns="x_filter">
      <rosparam>
        filter_field_name: x
        filter_limit_min: -3
        filter_limit_max: 3
        filter_limit_negative: false
      </rosparam>
    </group>
    <group ns="y_filter">
      <rosparam>
        filter_field_name: y
        filter_limit_min: -3
        filter_limit_max: 3
        filter_limit_negative: false
      </rosparam>
    </group>

    <!-- first, remove upper pointcloud -->
    <group ns="z_filter">
      <rosparam>
        filter_field_name: z
        filter_limit_negative: true
        filter_limit_min: 3.0
        filter_limit_max: 100.0
      </rosparam>
    </group>
    <group ns="odom_cloud">
      <rosparam>
        target_frame_id: odom
      </rosparam>
    </group>

  </group>
</launch>
